include/grpcpp/impl/codegen/proto_buffer_reader.h

```diff
@@ -0,0 +1,136 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPCPP_IMPL_CODEGEN_PROTO_BUFFER_READER_H
+#define GRPCPP_IMPL_CODEGEN_PROTO_BUFFER_READER_H
+
+#include <type_traits>
+
+#include <grpc/impl/codegen/byte_buffer_reader.h>
+#include <grpc/impl/codegen/grpc_types.h>
+#include <grpc/impl/codegen/slice.h>
+#include <grpcpp/impl/codegen/byte_buffer.h>
+#include <grpcpp/impl/codegen/config_protobuf.h>
+#include <grpcpp/impl/codegen/core_codegen_interface.h>
+#include <grpcpp/impl/codegen/serialization_traits.h>
+#include <grpcpp/impl/codegen/status.h>
+
+namespace grpc {
+
+extern CoreCodegenInterface* g_core_codegen_interface;
+
+/// An implementation of the ZeroCopyInputStream interface that uses
+/// \a grpc::ByteBuffer as the source of bytes from which to read.
+///
+/// Read more about ZeroCopyInputStream interface here:
+/// https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.io.zero_copy_stream#ZeroCopyInputStream
+///
+/// This is use to deserialize a contiguous stream of bytes into a Proto
+/// message. It may be inherited from to provide user application specific
+/// optimizations.
+class GrpcProtoBufferReader : public ::grpc::protobuf::io::ZeroCopyInputStream {
+ public:
+  /// Constructs buffer reader from \a buffer. Will set status to non ok if
+  /// \a buffer is invalid (the internal buffer has not been initialized).
+  explicit GrpcProtoBufferReader(ByteBuffer* buffer)
+      : byte_count_(0), backup_count_(0), status_() {
+    if (!buffer->Valid() ||
+        !g_core_codegen_interface->grpc_byte_buffer_reader_init(
+            &reader_, buffer->c_buffer())) {
+      status_ = Status(StatusCode::INTERNAL,
```

Comment:
This is the only place that `status_` looks to be set. It's not mutated after this that I can see. Thus, `status()` only appears to exist to signal construction failure. Would a factory for this make more sense?

```cpp
// returns nullptr on error
static std::unique_ptr<GrpcProtoBufferReader> GrpcProtoBufferReader::Reader(ByteBuffer*);
```

or some variant thereof that avoid the free store like

```cpp
static SomeEitherOrOptionalType<GrpcProtoBufferReader> GrpcProtoBufferReader::Make(ByteBuffer*);
```

Reply:
That makes sense to me. @vjpai, do you agree with this stylistically?